### PR TITLE
Improve KMS documentation

### PR DESCRIPTION
**Pull request by JoaoSa-da**

### CODE REVIEW

File: docs/2.8.0/docs/canton/usermanual/encrypted_private_key_storage.rst [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+There are two ways to choose the KMS wrapper key: (1) use an already existing KMS key or; (2)
+let Canton generate one.
+To use an already existing KMS key, you must specify its identifier. For example, for AWS KMS this can
+be one of the following:
+
+- Key id: `“1234abcd-12ab-34cd-56ef-1234567890ab”`
+- Key ARN (Amazon Resource Name): `“arn:aws:kms:us-east-1:1234abcd-12ab-34cd-56ef-1234567890ab”`
+- Key alias: `“alias/test-key”`
+
+Please be aware that an AWS KMS key needs to be configured with the following settings:
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/encrypted_private_key_storage.rst:86]: IMO we'd benefit from splitting things more and not mix GCP / AWS on the same page. We think of it as just different KMS providers, but an actual user will never use both AWS and GCP at the same time, so they don't need to have information for both on the same page. It forces them to filter out the stuff they need from what they don't care about.

I'd suggest we split this further into 4 pieces:

- KMS Envelope Encryption
  - AWS Setup guide
  - GCP Setup guide
- Full KMS Integration
  - AWS Setup guide
  - GCP Setup guide

We can keep a fifth page that links to these pages and provides the general concept explanation of what envelope vs full KMS is, benefits etc... like you already have.

If I put myself in the step of a user, what I'd like to know is, in order:
- What type of KMS encryption does Canton offer? Pros and Const (Envelope vs Full)
- What KMS vendors are supported? (GCP, AWS)
- How do I set up the combination of the above that I have chosen for my deployment

I think the docs should reflect more or less this flow

**Comment by thibault-da** [comment]
Added some comments for the first part, will do the second part today

File: docs/2.8.0/docs/canton/usermanual/kms/encrypted_private_key_storage_aws.rst [hunk 2]
```diff
@@ -0,0 +1,123 @@
+..
+   Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates.
+..
+   Proprietary code. All rights reserved.
+
+.. _encrypted_private_key_storage_aws:
+
+Configure Encrypted Private Key Storage with a AWS KMS
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/encrypted_private_key_storage_aws.rst:8]:
```suggestion
Configure Encrypted Private Key Storage with AWS KMS
```

File: docs/2.8.0/docs/canton/usermanual/kms/encrypted_private_key_storage_aws.rst [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+------------------------------
+
+To start using this feature we need to first enable AWS KMS for Canton:
+
+:ref:`--Configure AWS KMS for Canton-- <kms_aws_setup>`
+
+When using an AWS KMS to envelope encrypt the private keys stored in Canton, it needs to be configured
+with the following list of authorized actions (i.e. IAM permissions):
+
+- `kms:CreateKey`
```
[the remark this answers is not in the document]
Comment (thibault-da) [reply docs/2.8.0/docs/canton/usermanual/kms/encrypted_private_key_storage_aws.rst:31]: Ok, what I'm thinking is cross account key usage is probably less likely than same account usage. So I'd say that the permissions we put in this list should be for same account usage. Whatever other permissions are needed or not for cross account we can mention separately below.

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+Configure External Key Storage and Usage with a Key Management Service (KMS)
+============================================================================
+
+.. enterprise-only::
+
+Canton can use a KMS to `generate and store the private keys itself`.
+We directly use a KMS to generate and store Canton's private keys
+and then use its API to securely sign an decrypt messages. A Canton node still stores
+the corresponding public keys in its stores so that it can verify signatures and
+encrypt messages without having to rely on the KMS.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst:17]: It's not clear who `itself` refers to here. Is it Canton or the KMS? Maybe a better phrasing would be

```suggestion
Canton can delegate all asymmetric decryption and signing operations to a KMS. In that mode, private keys
are stored on the KMS, not in Canton's own storage nor in memory. Decryption and signing operations each
require an API call to the KMS. A Canton node still stores
the corresponding public keys in its stores so that it can verify signatures and
encrypt messages without having to rely on the KMS.
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+If you wish to know more you can read section: :ref:`Externalize Private Keys With a Key Management Service <kms_external_architecture>`.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst:19]:
```suggestion
See :ref:`Externalize Private Keys With a Key Management Service <kms_external_architecture>` for more details.
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst [hunk 6]
```diff
@@ -0,0 +1,23 @@
+..
+   Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates.
+..
+   Proprietary code. All rights reserved.
+
+.. _external_key_storage:
+
+Configure External Key Storage and Usage with a Key Management Service (KMS)
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst:8]: I think `Configure External Key Storage` is sufficient, WDYT? Seems implied that we're going to use them

```suggestion
Configure External Key Storage with a Key Management Service (KMS)
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+Canton can use a KMS to `generate and store the private keys itself`.
+We directly use a KMS to generate and store Canton's private keys
+and then use its API to securely sign an decrypt messages. A Canton node still stores
+the corresponding public keys in its stores so that it can verify signatures and
+encrypt messages without having to rely on the KMS.
+
+If you wish to know more you can read section: :ref:`Externalize Private Keys With a Key Management Service <kms_external_architecture>`.
+
+Currently Canton supports external key storage and usage with an `Amazon Web Services (AWS) KMS
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst:21]:
```suggestion
Currently Canton supports external key storage with `Amazon Web Services (AWS) KMS
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+<https://aws.amazon.com/kms/>`_ or
+a `Google Cloud Provider (GCP) KMS <https://cloud.google.com/security-key-management?hl=en>`_.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage.rst:23]:
```suggestion
`Google Cloud Provider (GCP) KMS <https://cloud.google.com/security-key-management?hl=en>`_.
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 9]
```diff
@@ -0,0 +1,199 @@
+..
+   Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates.
+..
+   Proprietary code. All rights reserved.
+
+.. _external_key_storage_aws:
+
+Configure External Key Storage and Usage with a AWS KMS
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:8]:
```suggestion
Configure External Key Storage with AWS KMS
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+=======================================================
+
+.. enterprise-only::
+
+The following section describes the steps needed to enable :ref:`External Key Storage and Usage <kms_external_architecture>`
+in Canton using AWS KMS. These steps include configuring AWS KMS, as well as, configuring this particular mode of operation.
+
+.. note::
+    Nodes configured with and without external KMS key storage can interact with each other,
+    including nodes using different KMS providers, as long as they follow the same crypto schemes.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:18]: This is true but `as long as they follow the same crypto schemes.` is a big enough caveat that we should emphasize it more. At least link it to the "Interoperability" section (see my other comment below)

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
+   :language: none
+   :start-after: user-manual-entry-begin: AwsKmsProviderFullConfig
+   :end-before: user-manual-entry-end: AwsKmsProviderFullConfig
+
+By default nodes run a ``tink`` provider that is not compatible with KMS provider. If you want to continue to have other nodes
+running a `non-KMS Canton` you are obliged to use a ``jce`` provider
+and you must explicitly configure it to use the KMS supported algorithms as the required algorithms. Here is an
+example for a domain:
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:67]: This paragraph assumes that the reader already has nodes running with KMS which is not necessarily the case. It's even at a bit at odds with the section which starts with
 
```
In the example below, we configure **a new** Canton participant node
```

This is of course important information but it's a bit confusing IMO that in the middle of the setup guide for AWS KMS on a participant I'm being told about tink being a default provider and that I need to configure jce on my domain. I'd put this in a separate sub-section, called something like "Interoperability with other nodes"

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+.. literalinclude:: /canton/includes/mirrored/enterprise/app/src/test/resources/aws-kms-provider-tagged.conf
+   :language: none
+   :start-after: user-manual-entry-begin: JceProviderParticipantConfig
+   :end-before: user-manual-entry-end: JceProviderParticipantConfig
+
+In other words, a node running with a ``kms`` provider is only ever able to communicate with other nodes running
+a ``kms`` or ``jce`` providers.
+
+A KMS only supports the :ref:`following cryptographic schemes <canton_supported_keys>`.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:84]: This links to a table detailing crypto schemes for all providers, not only KMS. So I'd rephrase to 

```suggestion
See :ref:`this table <canton_supported_keys>` for a description of cryptographic schemes supported by the KMS provider.
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+
+Setup with Pre-Generated Keys
+-----------------------------
+
+In the previous example, Canton creates its own keys on startup and initializes the identity of the nodes automatically.
+If we want to use pre-generated keys, we need to manually initialize the identity of the nodes by
+adding the following flag in the config:
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:91]:
```suggestion
adding the following to the node's configuration:
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+   :language: scala
+   :start-after: user-manual-entry-begin: ManualRegisterKmsKeys
+   :end-before: user-manual-entry-end: ManualRegisterKmsKeys
+   :dedent:
+
+where `xyzKmsKeyId` is the AWS KMS identifier for a specific key (e.g. `KMS Key RN`). If we are using, for example,
+`AWS cross account keys <https://docs.aws.amazon.com/kms/latest/developerguide/key-policy-modifying-external-accounts.html>`_
+be aware that using the key id is not enough and we are required to register the key using its `ARN`.
+
+Here is a brief reminder of the initial keys that each node owns:
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:112]:
```suggestion
Here is a reminder of the initial keys that each node owns:
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+| AWS               | - **Key Purpose:** `SIGN_VERIFY`                                      | - **Key Purpose:** `ENCRYPT_DECRYPT`                |
+|                   | - **Key Algorithms:** `ECC_NIST_P256` or `ECC_NIST_P384`              | - **Key Algorithm:** `RSA_2048`                     |
++-------------------+-----------------------------------------------------------------------+-----------------------------------------------------+
+
+Finally, using the previously registered keys, **we need to initialize our nodes**:
+
+- :ref:`initialize domain <manually-init-domain>`
+- :ref:`initialize participants <manually-init-participant>`
+
+Please be aware that for this specific use-case **we must use the keys that we registered and not generate new ones**.
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:139]: I'd put this before the links to the manual init doc. Because people will click on the link and possibly miss this and copy paste the script that generate new keys. So something like

```
Below are links to guides on how to manually initialize participant and domain nodes. In those guides, keys are generated using console commands such as:

val identityKey =
  participant.keys.secret.generate_signing_key(name = participant.name + "-namespace")

Make sure to replace those commands with the ones shown above with which we registered our existing keys instead of generating new ones.
```

File: docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+In other words, registering the keys replaces the `Keys Initialization` step in the guide to :ref:`manually initialize a node <manually_initializing_node>`.
+
+.. _participant_aws_kms_migration:
+
+Participant Node Migration to KMS Crypto Provider
```
Comment (thibault-da) [inline docs/2.8.0/docs/canton/usermanual/kms/external_key_storage_aws.rst:144]: This section and the one after (manual key rotation) seem to be exactly the same for AWS and GCP. I think those we can abstract away to their own page to avoid repeating

**Review by thibault-da — APPROVED** [review]

**Comment by thibault-da** [comment]
Thanks a lot @JoaoSa-da

File: docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+
+The authorized actions (i.e. IAM permissions) for the GCP KMS depend on the mode of operation we are running, either:
+(a) :ref:`an encrypted private key storage using envelope encryption <encrypted_private_key_storage_permissions_gcp>`; or
+(b) :ref:`externally storing and using keys directly from GCP KMS <external_key_storage_permissions_gcp>`.
+A complete list is given in those sections.
+
+For GCP, Canton uses a `GCP service account
+<https://cloud.google.com/docs/authentication/provide-credentials-adc#local-user-cred>`_. For example,
+the standard environment variable `GOOGLE_APPLICATION_CREDENTIALS` can be used after
+setting up a local Application Default Credentials (ADC) file for our service account.
```
Comment (carrielaben-da) [inline docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst:38]:
```suggestion
setting up a local Application Default Credentials (ADC) file for the service account.
```

File: docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+The protection and rotation of
+the credentials for GCP are the responsibility of the node operator.
+
+Auditability
+^^^^^^^^^^^^
+
+GCP provides tools to monitor KMS keys. For GCP you can refer to the `GCP official documentation
+<https://cloud.google.com/kms/docs/audit-logging>`_ for information on logging.
```
Comment (carrielaben-da) [inline docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst:47]:
```suggestion
GCP provides tools to monitor KMS keys. For GCP logging information you can refer to the `GCP official documentation
<https://cloud.google.com/kms/docs/audit-logging>`_.
```

File: docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+Errors resulting from the use of KMS keys are logged in Canton.
```
Comment (carrielaben-da) [inline docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst:48]:
```suggestion
Canton logs errors resulting from the use of KMS keys.
```

File: docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+
+Logging
+^^^^^^^
+
+For further auditability, Canton can be configured to log every call made to the GCP KMS.
+To enable this feature, set the ``audit-logging`` field of the KMS configuration to ``true``.
+By default, when using a file-based logging configuration, such logs will be written into the main canton log file.
```
Comment (carrielaben-da) [inline docs/2.8.0/docs/canton/usermanual/kms/kms_gcp_setup.rst:55]:
```suggestion
By default, when using a file-based logging configuration, such logs are written into the main Canton log file.
```

**Review by carrielaben-da — COMMENTED** [review]
Thanks!

**Review by carrielaben-da — APPROVED** [review]
LGTM!
